src/api/weth/WethApi.ts

```diff
@@ -0,0 +1,116 @@
+import Web3 from 'web3'
+
+import { Network, WithTxOptionalParams, Receipt } from 'types'
+import { WETH_ADDRESS_MAINNET, WETH_ADDRESS_RINKEBY } from 'const'
+import { wethAbi } from '@gnosis.pm/dex-js'
+import { logDebug } from 'utils'
+
+import { WethContract } from '@gnosis.pm/dex-js/build-esm/contracts/WethContract'
+
+export interface WrapUnwrapParams extends WithTxOptionalParams {
+  networkId: number
+  userAddress: string
+  amount: string
+}
+
+/**
+ * Interfaces for Wrapping ETH and unwrapping WETH token
+ */
+export interface WethApi {
+  deposit(params: WrapUnwrapParams): Promise<Receipt>
+  withdraw(params: WrapUnwrapParams): Promise<Receipt>
+}
+
+export interface WethApiDependencies {
+  web3: Web3
+  fetchGasPrice(): Promise<string | undefined>
+}
+
+function getWethAddressByNetwork(networkId: number): string {
+  switch (networkId) {
+    case Network.Mainnet:
+      return WETH_ADDRESS_MAINNET
+    case Network.Rinkeby:
+      return WETH_ADDRESS_RINKEBY
+    default:
+      throw new Error(`WethApi was not deployed to network ${networkId}`)
+  }
+}
+
+/**
+ * Basic implementation of WETH API
+ * Note that there's already another API for all ERC20, so only the deposit/withdraw is implemented in this API
+ */
+export class WethApiImpl implements WethApi {
+  protected _contractPrototype: WethContract
+  private web3: Web3
+  private fetchGasPrice: WethApiDependencies['fetchGasPrice']
+  protected static _contractsCache: { [network: number]: { [address: string]: WethContract } } = {}
+
+  public constructor(injectedDependencies: WethApiDependencies) {
+    Object.assign(this, injectedDependencies)
+    this._contractPrototype = new this.web3.eth.Contract(wethAbi) as WethContract
+  }
+
+  public async deposit(params: WrapUnwrapParams): Promise<Receipt> {
+    const { networkId, amount, userAddress, txOptionalParams } = params
+    const contract = await this._getContract(networkId)
+
+    const tx = contract.methods
+      .deposit()
+      .send({ from: userAddress, value: amount, gasPrice: await this.fetchGasPrice() })
+
+    if (txOptionalParams && txOptionalParams.onSentTransaction) {
+      tx.once('transactionHash', txOptionalParams.onSentTransaction)
+    }
+
+    logDebug(`[WethApi] Wrapped ${amount} ETH for user ${userAddress} in network ${networkId}`)
+
+    return tx
+  }
+
+  public async withdraw(params: WrapUnwrapParams): Promise<Receipt> {
+    const { networkId, amount, userAddress, txOptionalParams } = params
+    const contract = await this._getContract(networkId)
```

Comment:
same suggestion about putting both `async` fn calls inside 1 `await Promise.all([ ... ])`

Reply:
We don't do it for the other APIs. I don't think we should do it. If you really really feel is an issue, is not a big deal. But this becomes sync code, so no big point for moving those 2 promise together

Reply:
Discussed better in one place ☝️